### ISSUE TITLE
Warn if "ORDER BY" is used on multiple columns and only *some* columns specify explicit ordering

### ISSUE THREAD

**Issue by barrywhart** [sqlfluff/sqlfluff#419]
We had a bad bug recently in a project where somebody wrote this:

```
ORDER BY a, b DESC
```

They thought it meant this:
```
ORDER BY a DESC, b DESC
```

But it actually means this:
```
ORDER BY a ASC, b DESC
```

Suggestion: SQLFluff should warn that it's "potentially ambiguous" if an `ORDER BY` specifies the order direction `ASC` or `DESC` on _some_ but not _all_ columns. An automatic fix for this would be to have SQLFluff add `ASC` to any columns that didn't specify a sort order in the `ORDER BY`.

**Comment by alanmcruickshank** [comment]
one hundred times yes, this would be a GREAT rule! 💯 

I think taking inspiration from `Rule_L020` would be a good start to this one, although I think this rule could be much simpler that L020.

Who wants to take this?

**Comment by bdelamotte** [comment]
Hello! I'd like to try to take a stab at this. I really like sqlfluff. Niall said this would be a good first issue for me. I am hoping to carve out time this weekend to start on it if that's okay.

**Comment by bdelamotte** [comment]
Alright, I got a solution. It is currently working when I call sqlfluff.lint() on a string like so
```
    select a, b
      from faketable      
        ORDER BY a, b DESC 
```
Result:
```
 {'code': 'L035',
  'description': 'Ambiguous order by clause. Order by clauses should specify '
                 'order direction for each column in the order by clause.',
  'line_no': 4,
  'line_pos': 9},
```
I kept some bad formatting in the above code to make sure I saw other lint errors when I used the simple API.


However, I am getting the following when I try to call sqlfluff.fix() on the same code:
```
Loop limit on fixes reached [10]. Some fixes may be overdone.
('\n'
 '    select a, b\n'
 '      from faketable      \n'
 '        ORDER BY a ASC ASC ASC ASC ASC, b DESC \n')
```

I am unsure why this is happening but it could be how I'm tackling the problem. Could I push up a separate branch for this with the new code changes and get a quick review?

I will keep looking into why sqlfluff.fix() is giving this error this week some more.

Here is what I've added to `core/rules/std.py` on a local branch so far:
```

@std_rule_set.document_fix_compatible
@std_rule_set.register
class Rule_L035(BaseCrawler):
    """Ambiguous ordering directions for columns in order by clause.

    | **Anti-pattern**

    .. code-block::

        SELECT
            a, b
        FROM foo
        ORDER BY a, b DESC

    | **Best practice**
    | Specify ASC or DESC for all columns in the order by clause.

    .. code-block::

        SELECT
            a, b
        FROM foo
        ORDER BY a ASC, b DESC
    """

    def _eval(self, segment, parent_stack, **kwargs):
        """Ambiguous ordering directions for columns in order by clause.

        This rule checks for the ASC or DESC keyword for all columns in the order by clause.
        """
        # We only trigger on orderby_clause
        if segment.is_type("orderby_clause"):
            found_column_reference = False
            found_ordering_reference = False
            mismatch_found = False
            insert_buff = []
            lint_fixes = []
            for child_segment in segment.segments:
                if found_column_reference == True and found_ordering_reference == True:
                    # Reset findings
                    found_column_reference = False
                    found_ordering_reference = False
                if child_segment.is_type("column_reference"):
                    found_column_reference = True
                elif child_segment.is_type("keyword") and child_segment.name in ("ASC", "DESC"):
                    found_ordering_reference = True
                elif child_segment.type not in ["keyword", "whitespace", "indent", "dedent"]:
                    # Since ASC is default in SQL, add in ASC for fix
                    mismatch_found = True
                    new_whitespace = self.make_whitespace(raw=' ', pos_marker=child_segment.pos_marker)
                    new_keyword = self.make_keyword(raw="ASC", pos_marker=child_segment.pos_marker)
                    whitespace_lint_fix = LintFix("create", child_segment, new_whitespace)
                    order_lint_fix = LintFix("create", child_segment, new_keyword)
                    lint_fixes.append(whitespace_lint_fix)
                    lint_fixes.append(order_lint_fix)
                    insert_buff.append(
                        self.make_keyword(
                            raw="ASC", pos_marker=child_segment.pos_marker
                        )
                    )
                    found_ordering_reference = True # Here to cause reset

            if mismatch_found:
                # We've detected 1 or more columns that do not have the order direction
                return [
                    LintResult(
                        anchor=segment,
                        fixes=lint_fixes,
                        description=(
                            "Ambiguous order by clause. Order by "
                            "clauses should specify order direction for each column in the order by clause."
                        ),
                    )
                ]
            return None
        return None
```

The basic idea is to find a orderby_clause segment and look through its children. While looping through the children, try to find a column_reference segment. Once that is found, mark that one was found and look next for a keyword segment with name ASC or DESC. If we find any kind of indenting or other whitespace chars, we ignore, otherwise we throw an error. For example, if we find a comma segment after finding a column_reference but before a keyword segment with name ASC or DESC.

**Comment by barrywhart** [comment]
@bdelamotte: I just saw your note. Yes, please push up your branch. I will try and help with the issues you're seeing, and we can get others involved if necessary.

**Comment by barrywhart** [comment]
Assigned to myself. I'll begin by trying Brian's code.

**Comment by barrywhart** [comment]
I confirm the behavior Brian saw, where it's adding `ASC` multiple times. Next, I added some simple logging and verified that the `LintFix` / `ASC` is being applied multiple times. I'll see if I can determine why.

**Comment by barrywhart** [comment]
I found a bug in the fixing logic that causes `_eval()` to append `ASC` even if the column already specifies an order. Since SQLFluff has an outer loop that keeps running the rules on the output of the prior pass until no changes are produced or it hits a limit, this causes `ASC` to be appended several times. I tweaked the "reset" logic, and I think it's working correctly now. (Note: This is not fully tested.)

```
    def _eval(self, segment, parent_stack, **kwargs):
        """Ambiguous ordering directions for columns in order by clause.

        This rule checks for the ASC or DESC keyword for all columns in the order by clause.
        """
        # We only trigger on orderby_clause
        if segment.is_type("orderby_clause"):
            found_column_reference = False
            found_ordering_reference = False
            mismatch_found = False
            insert_buff = []
            lint_fixes = []
            for child_segment in segment.segments:
                if child_segment.is_type("column_reference"):
                    found_column_reference = True
                elif child_segment.is_type("keyword") and child_segment.name in ("ASC", "DESC"):
                    found_ordering_reference = True
                elif found_column_reference and child_segment.type not in ["keyword", "whitespace", "indent", "dedent"]:
                    if not found_ordering_reference:
                        # Since ASC is default in SQL, add in ASC for fix
                        mismatch_found = True
                        new_whitespace = self.make_whitespace(raw=' ', pos_marker=child_segment.pos_marker)
                        new_keyword = self.make_keyword(raw="ASC", pos_marker=child_segment.pos_marker)
                        whitespace_lint_fix = LintFix("create", child_segment, new_whitespace)
                        order_lint_fix = LintFix("create", child_segment, new_keyword)
                        lint_fixes.append(whitespace_lint_fix)
                        lint_fixes.append(order_lint_fix)
                        insert_buff.append(
                            self.make_keyword(
                                raw="ASC", pos_marker=child_segment.pos_marker
                            )
                        )
                    # Reset findings
                    found_column_reference = False
                    found_ordering_reference = False

            if mismatch_found:
                # We've detected 1 or more columns that do not have the order direction
                return [
                    LintResult(
                        anchor=segment,
                        fixes=lint_fixes,
                        description=(
                            "Ambiguous order by clause. Order by "
                            "clauses should specify order direction for each column in the order by clause."
                        ),
                    )
                ]
            return None
        return None
```

@bdelamotte: Do you want to continue the work on this? I think the next step is to add some tests to ensure it works correctly in a variety of cases. I'll unassign myself from the issue for now.

**Comment by barrywhart** [comment]
Correction -- I suggest one change to the "fix" -- I propose adding `ASC` to unspecified columns **if and only if** some columns **do** specify `ASC` or `DESC`. If **no** column explicitly specifies the sort order, do not make any changes. The intent is to correct ambiguities, and in my opinion, it's only ambiguous when there's a mix of specified and unspecified columns.

**Comment by bdelamotte** [comment]
@barrywhart , sorry for the delay in response. I was on paternity leave and then my email notifications got lost. I am glad you got this working and thanks for running with what I had!

**Comment by barrywhart** [comment]
@bdelamotte: Thank you so much for the code! This was the first time I'd worked on a lint rule that made fixes to the code, and your code got me 80% of the way to finished. 💯
